web/packages/teleport/src/UnifiedResources/Resources.tsx

```diff
@@ -50,56 +53,39 @@ export function Resources() {
     fieldName: 'name',
     dir: 'ASC',
   });
-  const {
-    params,
-    search,
-    setParams,
-    replaceHistory,
-    pathname,
-    setSort,
-    onLabelClick,
-  } = filtering;
-
-  const { fetchInitial, fetchedData, attempt, fetchMore } = useInfiniteScroll({
+  const { params, setParams, replaceHistory, pathname, setSort, onLabelClick } =
+    filtering;
+
+  const { fetch, forceFetch, resources, attempt } = useKeyBasedPagination({
     fetchFunc: teleCtx.resourceService.fetchUnifiedResources,
     clusterId,
-    params,
+    filter: params,
   });
 
-  useEffect(() => {
-    fetchInitial();
-  }, [clusterId, search]);
-
-  const infiniteScrollDetector = useRef(null);
-
-  // Install the infinite scroll intersection observer.
-  //
-  // TODO(bl-nero): There's a known issue here. We need to have `fetchMore` in
-  // the list of hook dependencies, because using a stale `fetchMore` closure
-  // means we will fetch the same data over and over. However, as it's
-  // implemented now, every time `fetchMore` changes, we reinstall the observer.
-  // This is mitigated by `fetchMore` implementation, which doesn't spawn
-  // another request before the first one finishes, but it's still a potential
-  // for trouble in future. We need to decouple updating the `fetchMore` closure
-  // and installing the observer.
-  useEffect(() => {
-    if (infiniteScrollDetector.current) {
-      const observer = new IntersectionObserver(entries => {
-        if (entries[0]?.isIntersecting) {
-          fetchMore();
-        }
-      });
-      observer.observe(infiniteScrollDetector.current);
-      return () => observer.disconnect();
-    }
-  });
+  const scrollDetector = useRef(null);
+
+  useInfiniteScroll(scrollDetector.current, fetch);
```

Comment:
`useInfiniteScroll` passes `fetch` as one of the deps to `useEffect`, but `fetch` has a new identity on each new render. Meaning the `useEffect` in `useInfiniteScroll` is called on each render, 7 times when just rendering the page and I'm not sure if it interferes with the actual scroll or not.

I tried to refactor it with no luck, I still get 7 console logs when opening the resources page.

<details>
<summary>Patch</summary>

```diff
diff --git a/web/packages/teleport/src/components/hooks/useInfiniteScroll.ts b/web/packages/teleport/src/components/hooks/useInfiniteScroll.ts
index 00cf2711cb..2ab6cd5ae3 100644
--- a/web/packages/teleport/src/components/hooks/useInfiniteScroll.ts
+++ b/web/packages/teleport/src/components/hooks/useInfiniteScroll.ts
@@ -42,6 +42,7 @@ export function useInfiniteScroll(trigger: Element, callback: () => void) {
     if (observer.current) {
       observer.current.disconnect();
     }
+    console.log('Creating a new observer.');
     observer.current = new IntersectionObserver(entries => {
       if (entries[0]?.isIntersecting) {
         callback();
diff --git a/web/packages/teleport/src/components/hooks/useKeyBasedPagination.ts b/web/packages/teleport/src/components/hooks/useKeyBasedPagination.ts
index 78b749e7c0..be38c44af0 100644
--- a/web/packages/teleport/src/components/hooks/useKeyBasedPagination.ts
+++ b/web/packages/teleport/src/components/hooks/useKeyBasedPagination.ts
@@ -14,7 +14,7 @@
  * limitations under the License.
  */
 
-import { useState, useEffect, useRef } from 'react';
+import { useCallback, useState, useEffect, useRef } from 'react';
 import useAttempt, { Attempt } from 'shared/hooks/useAttemptNext';
 
 import {
@@ -58,52 +58,66 @@ export function useKeyBasedPagination<T>({
 
   // This is an additional countermeasure to prevent fetching if `fetch` is
   // called multiple times per React state reconciliation cycle.
-  let fetchCalled = false;
-  const fetch = async (force: boolean) => {
-    if (
-      finished ||
-      fetchCalled ||
-      (!force &&
-        (attempt.status === 'processing' || attempt.status === 'failed'))
-    ) {
-      return;
-    }
-
-    try {
-      fetchCalled = true;
-      setAttempt({ status: 'processing' });
-      const limit = resources.length > 0 ? fetchMoreSize : initialFetchSize;
-      abortController.current?.abort();
-      abortController.current = new AbortController();
-      const res = await fetchFunc(
-        clusterId,
-        {
-          ...filter,
-          limit,
-          startKey,
-        },
-        abortController.current.signal
-      );
-      abortController.current = null;
-      setResources([...resources, ...res.agents]);
-      setStartKey(res.startKey);
-      if (!res.startKey) {
-        setFinished(true);
-      }
-      setAttempt({ status: 'success' });
-    } catch (err) {
-      // Aborting is not really an error here.
-      if (isAbortError(err)) {
-        setAttempt({ status: '', statusText: '' });
+  const fetchCalled = useRef(false);
+  const fetch = useCallback(
+    async (force: boolean) => {
+      if (
+        finished ||
+        fetchCalled.current ||
+        (!force &&
+          (attempt.status === 'processing' || attempt.status === 'failed'))
+      ) {
         return;
       }
-      setAttempt({ status: 'failed', statusText: err.message });
-    }
-  };
+
+      try {
+        fetchCalled.current = true;
+        setAttempt({ status: 'processing' });
+        const limit = resources.length > 0 ? fetchMoreSize : initialFetchSize;
+        abortController.current?.abort();
+        abortController.current = new AbortController();
+        const res = await fetchFunc(
+          clusterId,
+          {
+            ...filter,
+            limit,
+            startKey,
+          },
+          abortController.current.signal
+        );
+        abortController.current = null;
+        setResources(resources => [...resources, ...res.agents]);
+        setStartKey(res.startKey);
+        if (!res.startKey) {
+          setFinished(true);
+        }
+        setAttempt({ status: 'success' });
+      } catch (err) {
+        // Aborting is not really an error here.
+        if (isAbortError(err)) {
+          setAttempt({ status: '', statusText: '' });
+          return;
+        }
+        setAttempt({ status: 'failed', statusText: err.message });
+      }
+    },
+    [
+      attempt.status,
+      clusterId,
+      fetchFunc,
+      fetchMoreSize,
+      filter,
+      finished,
+      initialFetchSize,
+      setAttempt,
+      startKey,
+      resources.length,
+    ]
+  );
 
   return {
-    fetch: () => fetch(false),
-    forceFetch: () => fetch(true),
+    fetch: useCallback(() => fetch(false), [fetch]),
+    forceFetch: useCallback(() => fetch(true), [fetch]),
     attempt,
     resources,
     finished,

```
</details>

One of the deps of `fetch` in `useKeyBasedPagination` must be getting recreated on each render.

Reply:
I'd mostly just check if it doesn't run each time you scroll or something like this. I didn't check this as I was too lazy to do this trick where the backend returns dozens of resources.